File: todolist/src/server/service/weather/weatherNmcService.js
const {WEATHER_HEADERS} = require("../../utils/constants");

const fetch = require('node-fetch');

// 格式化NMC数据为统一格式
function extractNmcWeatherData(nmcData) {
    if (!(nmcData && nmcData.code === 0 && nmcData.data)) {
        return {};
    }
    const weatherData = {
        liveWeather: {},    // 实况天气
        dailyWeather: [],   // 近几天预报
    };
    if (nmcData.data.predict?.detail && Array.isArray(nmcData.data.predict.detail)) {
        weatherData.dailyWeather = nmcData.data.predict.detail.map(item => ({
            date: item.date?.trim().replaceAll('-','') || '',
            weather: (item.day?.weather?.info || '').replace('9999', ''),
            tempMin: (item.night?.weather?.temperature || '').replace('9999', ''),
            tempMax: (item.day?.weather?.temperature || '').replace('9999', ''),
            wind: `${item.day?.wind?.direct || ''} ${item.day?.wind?.power || ''}`.replaceAll('9999', '').trim(), // 风向风力
        }));
    }
    weatherData.liveWeather = {
        time: nmcData.data.real?.publish_time?.trim().split(' ')[1] || '', // 从publish_time获取时间，只保留HH:MM格式
        weather: nmcData.data.real?.weather?.info || '',
        temperature: nmcData.data.real?.weather?.temperature || '',
        tempMax: weatherData.dailyWeather?.[0]?.tempMax || '',
        tempMin: weatherData.dailyWeather?.[0]?.tempMin || '',
        wind: `${nmcData.data.real?.wind?.direct || ''} ${nmcData.data.real?.wind?.power || ''}`, // 风向风力
        humidity: nmcData.data.real?.weather?.humidity || '', // 湿度
        airQuality: `${nmcData.data.air?.aqi} ${nmcData.data.air?.text || ''}`, // 空气质量
        sunrise: nmcData.data.real?.sunriseSunset?.sunrise?.trim().split(' ')[1] || '', // 日出
        sunset: nmcData.data.real?.sunriseSunset?.sunset?.trim().split(' ')[1] || '', // 日落
    };
    return weatherData;
}

// 获取中央气象台天气数据
async function fetchNmcWeather(nmcApiCode){
    if (!nmcApiCode) {
        return { error: { message: '参数nmcApiCode为空，无法获取中央气象台天气数据' } };
    }
    try {
        // 仅使用指定的URL接口，使用当前时间戳
        const nmcWeatherUrl = `https://www.nmc.cn/rest/weather?stationid=${nmcApiCode}&_=${Date.now()}`;
        console.log(`开始获取中央气象台天气数据，正在访问: ${nmcWeatherUrl}`);
        const weatherResponse = await fetch(nmcWeatherUrl, { method: 'GET', headers: WEATHER_HEADERS, timeout: 5000 });

        if (!weatherResponse.ok) {
            throw new Error(`HTTP响应状态码: ${weatherResponse.status}`);
        }

        const nmcData = await weatherResponse.json();
        const nmcWeatherData = extractNmcWeatherData(nmcData);
        console.log('成功提取中央气象台天气数据');
        return nmcWeatherData;
    } catch (error) {
        console.error('获取中央气象台天气数据失败:', error);
        return { error: { message: error.message || '获取中央气象台天气数据失败' } };
    }
}

module.exports = {
    fetchNmcWeather
};

File: todolist-python/app/services/weather/weather_nmc_service.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
中央气象台天气服务模块

提供从中央气象台API获取天气数据的功能，与Node.js项目中的weatherNmcService.js功能对应。
"""

import requests
import time
from app.utils.logger_util import logger


# 天气请求头
WEATHER_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
    "Connection": "keep-alive",
    "Referer": "https://www.nmc.cn/"
}


def extract_nmc_weather_data(nmc_data):
    """
    格式化NMC数据为统一格式
    
    Args:
        nmc_data: 中央气象台返回的原始数据
    
    Returns:
        格式化后的天气数据
    """
    if not (nmc_data and nmc_data.get('code') == 0 and nmc_data.get('data')):
        return {}
    
    weather_data = {
        "liveWeather": {},    # 实况天气
        "dailyWeather": [],   # 近几天预报
    }
    
    data = nmc_data.get('data')
    
    # 提取近几天预报
    if data.get('predict') and data.get('predict').get('detail') and isinstance(data.get('predict').get('detail'), list):
        weather_data['dailyWeather'] = [
            {
                "date": item.get('date', '').strip().replace('-', '') if item.get('date') else '',
                "weather": (item.get('day', {}).get('weather', {}).get('info', '')).replace('9999', ''),
                "tempMin": (item.get('night', {}).get('weather', {}).get('temperature', '')).replace('9999', ''),
                "tempMax": (item.get('day', {}).get('weather', {}).get('temperature', '')).replace('9999', ''),
                "wind": f"{item.get('day', {}).get('wind', {}).get('direct', '')} {item.get('day', {}).get('wind', {}).get('power', '')}".replace('9999', '').strip()
            }
            for item in data.get('predict').get('detail')
        ]
    
    # 提取实况天气
    real = data.get('real', {})
    publish_time = real.get('publish_time', '')
    time_str = publish_time.strip().split(' ')[1] if publish_time else ''
    
    weather = real.get('weather', {}).get('info', '')
    temperature = real.get('weather', {}).get('temperature', '')
    wind_direct = real.get('wind', {}).get('direct', '')
    wind_power = real.get('wind', {}).get('power', '')
    humidity = real.get('weather', {}).get('humidity', '')
    
    # 获取空气质量
    air_aqi = data.get('air', {}).get('aqi', '')
    air_text = data.get('air', {}).get('text', '')
    air_quality = f"{air_aqi} {air_text}" if air_text else f"{air_aqi}"
    
    # 获取日出日落时间
    sunrise = ''
    sunset = ''
    if real.get('sunriseSunset'):
        sunrise_time = real.get('sunriseSunset').get('sunrise', '')
        sunset_time = real.get('sunriseSunset').get('sunset', '')
        sunrise = sunrise_time.strip().split(' ')[1] if sunrise_time else ''
        sunset = sunset_time.strip().split(' ')[1] if sunset_time else ''
    
    # 获取最高最低温度
    temp_max = ''
    temp_min = ''
    if weather_data['dailyWeather']:
        temp_max = weather_data['dailyWeather'][0].get('tempMax', '')
        temp_min = weather_data['dailyWeather'][0].get('tempMin', '')
    
    weather_data['liveWeather'] = {
        "time": time_str,
        "weather": weather,
        "temperature": temperature,
        "tempMax": temp_max,
        "tempMin": temp_min,
        "wind": f"{wind_direct} {wind_power}",
        "humidity": humidity,
        "airQuality": air_quality,
        "sunrise": sunrise,
        "sunset": sunset
    }
    
    return weather_data


def fetch_nmc_weather(nmc_api_code):
    """
    获取中央气象台天气数据
    
    Args:
        nmc_api_code: 中央气象台API编码
    
    Returns:
        天气数据或错误信息
    """
    if not nmc_api_code:
        return {"error": {"message": "参数nmc_api_code为空，无法获取中央气象台天气数据"}}
    
    try:
        # 仅使用指定的URL接口，使用当前时间戳
        timestamp = int(time.time() * 1000)
        nmc_weather_url = f"https://www.nmc.cn/rest/weather?stationid={nmc_api_code}&_={timestamp}"
        logger.info(f"开始获取中央气象台天气数据，正在访问: {nmc_weather_url}")
        
        response = requests.get(nmc_weather_url, headers=WEATHER_HEADERS, timeout=5)
        
        if not response.status_code == 200:
            raise Exception(f"HTTP响应状态码: {response.status_code}")
        
        nmc_data = response.json()
        nmc_weather_data = extract_nmc_weather_data(nmc_data)
        logger.info("成功提取中央气象台天气数据")
        
        return nmc_weather_data
    except Exception as e:
        logger.error(f"获取中央气象台天气数据失败: {e}")
        return {"error": {"message": str(e) or "获取中央气象台天气数据失败"}}
